web/packages/teleport/src/JoinTokens/CreateJoinTokenDialog.tsx

```diff
@@ -0,0 +1,565 @@
+            placeholder="AWS ARN"
+            value={rule.aws_arn}
+            onChange={e => setTokenRulesField(index, 'aws_arn', e.target.value)}
+          />
+        </RuleBox>
+      ))}
+      <ButtonText onClick={addNewRule}>
+        <Plus size={16} mr={2} />
+        Add another AWS Rule
+      </ButtonText>
+    </>
+  );
+};
+
+type OptionGCP = Option<string, string>;
+
+const JoinTokenGCPForm = ({
+  tokenState,
+  onUpdateState,
+}: {
+  tokenState: NewJoinTokenState;
+  onUpdateState: (newToken: NewJoinTokenState) => void;
+}) => {
+  const rules = tokenState.gcp;
+  function removeRule(index: number) {
+    const newRules = rules.filter((_, i) => index !== i);
+    const newState = {
+      ...tokenState,
+      gcp: newRules,
+    };
+    onUpdateState(newState);
+  }
+
+  function addNewRule() {
+    const newState = {
+      ...tokenState,
+      gcp: [
+        ...tokenState.gcp,
+        { project_ids: [], locations: [], service_accounts: [] },
+      ],
+    };
+    onUpdateState(newState);
+  }
+
+  function updateRuleField(
+    index: number,
+    fieldName: string,
+    opts: OptionGCP[]
+  ) {
+    const newState = {
+      ...tokenState,
+      gcp: tokenState.gcp.map((rule, i) => {
+        if (i === index) {
+          return { ...rule, [fieldName]: opts };
+        }
+        return rule;
+      }),
+    };
+    onUpdateState(newState);
+  }
+
+  return (
+    <>
+      {rules.map((rule, index) => (
+        <RuleBox
+          key={index} // order doesn't change without updating the referrenced array
+        >
+          <Flex alignItems="center" justifyContent="space-between">
+            <Text fontWeight={700} mb={2}>
+              GCP Rule
+            </Text>
+
+            {rules.length > 1 && ( // at least one rule is required, so lets not allow the user to remove it
+              <ButtonIcon onClick={() => removeRule(index)}>
+                <Trash size={16} color="text.muted" />
+              </ButtonIcon>
+            )}
+          </Flex>
+          <FieldSelectCreatable
+            placeholder="Type a project ID"
+            isMulti
+            isClearable
+            isSearchable
+            onChange={opts =>
+              updateRuleField(index, 'project_ids', opts as OptionGCP[])
+            }
+            value={rule.project_ids}
+            label="Add Project ID(s)"
+            rule={requiredField('At least 1 Project ID required')}
+          />
+          <FieldSelectCreatable
+            placeholder="us-west1, us-east1-a"
+            isMulti
+            isClearable
+            isSearchable
+            onChange={opts =>
+              updateRuleField(index, 'locations', opts as OptionGCP[])
+            }
+            value={rule.locations}
+            label="Add Locations"
+            labelTip="Allows regions and/or zones."
+          />
+          <FieldSelectCreatable
+            placeholder="PROJECT_NUMBER_compute@developer.gserviceaccount.com"
```

Comment:
maybe a shorter email example? b/c i can't see the entire text in the input box

Reply:
I think its in a specific format according to the docs, but i'll remove `_NUMBER` and that'll shorten it enough i think